test/jdk/sun/net/www/http/HttpClient/MultiThreadTest.java

```diff
@@ -139,61 +177,98 @@ public static void main(String args[]) throws Exception {
+                if (w.headerFound > 0) {
+                    validConnections++;
+                    if (w.mayHaveTimedOut(end)) {
+                        debug("Worker " + w.id + " may have timed out");
+                        becameIdle++;
+                    } else {
+                        long at0 = at(w.lastReading);
+                        long at1 = at(w.lastReplied);
+                        debug("Worker " + w.id +" has not timed out - last used at " +
+                            Math.max(at0, at1));
+                    }
+                } else {
+                    debug("Worker " + w.id + " is not a valid connection");
+                }
+            }
 
-        if (validConnections > threads + 1 || validConnections == 0) { // could be less
-            throw new RuntimeException ("Expected " + threads + " connections: used " + validConnections);
-        }
+            if (validConnections > threads) {
+                if (SLEEP > KEEP_ALIVE) {
+                    debug("INFO: " + validConnections
+                            + " have been used, with " + becameIdle
+                            + " becoming idle for more than " + KEEP_ALIVE + "s"
+                            + " while using " + threads
+                            + " threads to make concurrent connections");
+                } else {
+                    debug("WARNING: " + validConnections
+                            + " have been used, with " + becameIdle
+                            + " becoming idle for more than " + KEEP_ALIVE + "s"
+                            + " where only " + threads
+                            + " connections and none idle were expected!");
+                }
+            }
 
-        // Sometimes the client drops a connection after a while and
-        // spawns a new one. Why this is happening is not clear,
-        // and JDK-8223783 is logged to follow up on this. For the sake
-        // of test stabilization we don't fail on `threads + 1` connections
-        // but log a warning instead.
-        if (validConnections == threads + 1) {
-            debug("WARNING: " + validConnections
-                + " have been used, where only " + threads
-                + " were expected!");
-        }
+            if (validConnections > threads + becameIdle || validConnections == 0) { // could be less
+                throw new RuntimeException("Expected " + (threads + becameIdle) + " connections: used " + validConnections);
+            }
 
-        if (validConnections != cnt) {
-            debug("WARNING: got " + (cnt - validConnections) + " unexpected connections!");
-        }
-        if  (validConnections == cnt && reqs != threads*requests) {
-            throw new RuntimeException ("Expected "+ threads*requests+ " requests: got " +reqs);
-        }
+            if (validConnections != cnt) {
+                // these connections are not necessarily unexpected if SLEEP exceeds keep-alive.
```

Comment:
Well actually these are unexpected connections; every connection from the tested client will equally count towards `cnt` and `validConnections`.

Reply:
Ah! good point. I mixed up cnt and threads. I will clean that up.